packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1457,92 +1557,334 @@ class Navigator extends StatefulWidget {
+    doingPop = false;
+  }
+
+  bool _reportRemovalToObserver = true;
+
+  // Route is removed without being completed.
+  void remove({ bool isReplaced = false }) {
+    if (currentState.index >= _RouteLifecycle.remove.index)
+      return;
+    assert(isPresent);
+    _reportRemovalToObserver = !isReplaced;
+    currentState = _RouteLifecycle.remove;
+  }
+
+  // Route completes with `result` and is removed.
+  void complete<T>(T result, { bool isReplaced = false }) {
+    if (currentState.index >= _RouteLifecycle.remove.index)
+      return;
+    assert(isPresent);
+    _reportRemovalToObserver = !isReplaced;
+    route.didComplete(result);
+    assert(route._popCompleter.isCompleted); // implies didComplete was called
+    currentState = _RouteLifecycle.remove;
+  }
+
+  void finalize() {
+    assert(currentState.index < _RouteLifecycle.dispose.index);
+    currentState = _RouteLifecycle.dispose;
+  }
+
+  void dispose() {
+    assert(currentState.index < _RouteLifecycle.disposed.index);
+    route.dispose();
+    currentState = _RouteLifecycle.disposed;
+  }
+
+  bool get isPresent => currentState.index <= _RouteLifecycle.idle.index;
+
+  bool shouldAnnounceChangeToNext(Route<dynamic> nextRoute) {
+    assert(nextRoute != lastAnnouncedNextRoute);
+    // Do not announce if `next` changes from a just popped route to null. We
+    // already announced this change by calling didPopNext.
+    return !(
+      nextRoute == null &&
+      lastAnnouncedPoppedNextRoute != null &&
+      lastAnnouncedPoppedNextRoute == lastAnnouncedNextRoute
+    );
+  }
+
+  static final _RouteEntryPredicate isPresentPredicate = (_RouteEntry entry) => entry.isPresent;
+
+  static _RouteEntryPredicate isRoutePredicate(Route<dynamic> route) {
+    return (_RouteEntry entry) => entry.route == route;
+  }
+}
+
+/// The state for a [Navigator] widget.
+class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+  final List<_RouteEntry> _history = <_RouteEntry>[];
+
+  bool _debugLocked = false; // used to prevent re-entrant calls to push, pop, and friends
+
+  /// The [FocusScopeNode] for the [FocusScope] that encloses the routes.
+  final FocusScopeNode focusScopeNode = FocusScopeNode();
+
+  /// The overlay this navigator uses for its visual presentation.
+  OverlayState get overlay => _overlayKey.currentState;
+  final GlobalKey<OverlayState> _overlayKey = GlobalKey<OverlayState>();
+
+  Iterable<OverlayEntry> get _allRouteOverlayEntries sync* {
+    for (_RouteEntry entry in _history)
+      yield* entry.route.overlayEntries;
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    for (NavigatorObserver observer in widget.observers) {
+      assert(observer.navigator == null);
+      observer._navigator = this;
+    }
+    // TODO(goderbauer): don't use defaultRouteName if pages are provided
+    _history.addAll(
+      widget.onGenerateInitialRoutes(this, widget.initialRoute ?? Navigator.defaultRouteName)
```

Comment:
I'm not sure this will work with our `setInitialRoute` logic in the embedders.  They expect to be able to set that and have it piped through to `window.defaultRouteName`, whereas this looks like it will never receive that value (unless I'm missing something).

Reply:
It was (and still is after this change) the responsibility of the WidgetApp to provide the navigator with the initialRoute taken from the Window: https://github.com/flutter/flutter/blob/387f885481cd8441237680de0cfcd0802c0af48c/packages/flutter/lib/src/widgets/app.dart#L1091

Let me know if you see any problems with this.

Reply:
Ahhh that's right.  Ok.

Reply:
Ahhh that's right.  Ok.